Proxy protocol support for SMTP and IMAP

drdaeman opened the pull request:
Hey, first of all, once again thank you for this project! I need to put maddy behind a reverse proxy (Traefik, but shouldn't really matter), but it needs to be aware about the original IP address, so I want it to support PROXY protocol.

What do you think about this implementation? I haven't fully battle-tested it yet, but it seem to work and to best of my awareness it should resolve two out of three remaining point of #369. Of course, I want this to be accepted, so I'm willing to make this PR as good as possible (if needed) and would appreciate any feedback.

Two potential issues I see are related to tests:
- First is that I've (ab)used SPF to test that the proxied IP address is recognized, as I'm not aware of any way to get it back. So PROXY tests assume that SPF is working correctly, which isn't exactly great. Your advice on this is most appreciated - maybe there is some debug command that I can issue to get test SMTP server to echo who it thinks is connecting?
- Another issue is that I cannot get tests to pass reliably. Even running one single test (`./run.sh -run TestProxyProtocolTrusted`) is flaky and half of the time results with a nondescript failure (all logs seem to be fine but it ends with "t.go:294: The server did not stop cleanly, deadlock?"). And when running multiple tests all together it seems that there's some interference between them and that trusted address list from one TestProxyProtocol function can somehow leak into another. I haven't investigated this further as it seems to be a different issue and could be just something about my local Go installation.

Thanks!

drdaeman committed:
Proxy protocol support for SMTP and IMAP

foxcpp committed:
Bump SPF library version

From library changelog:

mx: Continue on DNS response with invalid names
      On MX lookups, if there is one invalid MX record (for example, pointing
to an IP address instead of a name), Go will return an error and also
the valid records.

Today, we just fail this lookup as if we had a DNS error.

However, the SPF is not clear about this specific case, and all other
popular implementations will just ignore the invalid record and keep
processing using the valid ones.

This patch makes us match common practice.

arush15june committed:
feat: add iam, aws config file, minio config file configurable credentials for s3 blob storage

thohui committed:
fix: typo

foxcpp committed:
Merge pull request #544 from thohui/fix-typo

fix: typo

Commit:
doc: replace `maddyctl` with `maddy`

Commit:
fixes #546 - ldap hang

foxcpp committed:
Merge pull request #547 from angelnu/fix_ldap_hang

fixes #546 - ldap hang

foxcpp committed:
Merge pull request #537 from arush15june/feat-s3-creds

internal/storage/blob/s3: add iam, aws config file, minio config file configurable credentials

foxcpp committed:
Merge pull request #545 from pouriya/remove-maddyctl-from-docs

doc: replace `maddyctl` with `maddy`

domcyrus committed:
fix race condition on table.file #557

domcyrus committed:
improve file test a bit further

foxcpp committed:
Merge pull request #559 from domcyrus/fix-race-on-table-file

fix race condition on table.file #557

cfbraun committed:
fix: allow empty From address to match with TakeMsg

foxcpp committed:
Merge pull request #561 from cfbraun/master

fix: allow empty From address to match with TakeMsg

foxcpp committed:
Improve domain validation function

Now FQDNs are accepted as well (with trailing dot).
Empty strings are not considered valid.
Label length for IDNs is checked using Punycode form as it should.

See https://github.com/foxcpp/maddy/issues/554

foxcpp committed:
Bump go-imap-mess version

foxcpp committed:
Bump libdns implementations versions to dev versions

Done to make sure go.mod references commits rather than tags.
There is some evidence libdns developers overwrite them, causing
go.sum failures when building maddy. The only known offenders
are libdns/alidns and libdns/gandi, but I did the change for all
libdns implementations just in case.

See #461.

foxcpp committed:
Bump required Go version to 1.18

foxcpp committed:
Bump go-imap-sql version

See #549.

arisudesu committed:
Extract mapping lookup from loop

arisudesu committed:
Fix undefined variable err

foxcpp committed:
docs: Improve authorize_sender documentation

Add prepare_email description. Expand user_to_email description.

See #565.

foxcpp committed:
Merge branch 'dev'

foxcpp committed:
target: Fix tests to use support go-smtp

foxcpp committed:
module: Fix ModSpecificData initialization

foxcpp committed:
go get -u ./... && go mod tidy

Except github.com/lib/pq, it stays on 1.10.6 since 1.10.7
seems to introduce a breaking change.

foxcpp committed:
maddy 0.7.0

the-maldridge committed:
docs/reference/auth: Add information about netauth support

drdaeman commented on tests/smtp_test.go:
```diff
@@ -68,6 +69,94 @@ func TestCheckRequireTLS(tt *testing.T) {
+			tls off
+
+			proxy_protocol {
+				trust ` + tests.DefaultSourceIP.String() + ` ::1/128
```
IPv6 address here is to ensure parser recognizes both IPv4 (presuming `DefaultSourceIP` is IPv4) and IPv6.

drdaeman commented on tests/smtp_test.go:
```diff
@@ -68,6 +69,94 @@ func TestCheckRequireTLS(tt *testing.T) {
+
+	conn := t.Conn("smtp")
+	defer conn.Close()
+	conn.Writeln(fmt.Sprintf("PROXY TCP4 192.168.99.1 %s 12345 %d", tests.DefaultSourceIP.String(), t.Port("smtp")))
```
Hardcoded IP address may be not exactly a best approach. Is this fine as it is, or should I add another constant (`SecondarySourceIP`)?

foxcpp replied:
Probably using `127.0.0.17` would be better to avoid any traffic if test misbehaves and actually tries to contact 192.168.99.1.

drdaeman commented on tests/smtp_test.go:
```diff
@@ -68,6 +69,94 @@ func TestCheckRequireTLS(tt *testing.T) {
+	conn.Writeln(fmt.Sprintf("PROXY TCP4 192.168.99.1 %s 12345 %d", tests.DefaultSourceIP.String(), t.Port("smtp")))
+	conn.SMTPNegotation("localhost", nil, nil)
+	conn.Writeln("MAIL FROM:<testing@one.maddy.test>")
+	conn.ExpectPattern("250 *")
```
As I've mentioned in the PR description, I'm abusing SPF checker to ensure that SMTPd thinks it serves a client from 192.168.99.1 and not 127.109.97.100. It surely works, although I'm not really happy about this and if someone can advise me a different approach I would be more than happy to try something else.

drdaeman commented:
I've battle-tested this on my own SMTP service running latest Maddy with my patches included for the last week, using `proxy_protocol` behind Traefik - and it seem to work without issues. All the headers and logs are spot on, no 192.168s. I haven't production-tested IMAP - I run Dovecot (because of dsync) - but I don't suppose it should be any different.

dependabot[bot] committed:
build(deps): bump golang.org/x/net from 0.5.0 to 0.7.0

Bumps [golang.org/x/net](https://github.com/golang/net) from 0.5.0 to 0.7.0.
- [Release notes](https://github.com/golang/net/releases)
- [Commits](https://github.com/golang/net/compare/v0.5.0...v0.7.0)

---
updated-dependencies:
- dependency-name: golang.org/x/net
  dependency-type: direct:production
...

Signed-off-by: dependabot[bot] <support@github.com>

foxcpp committed:
Merge pull request #573 from foxcpp/dependabot/go_modules/golang.org/x/net-0.7.0

build(deps): bump golang.org/x/net from 0.5.0 to 0.7.0

foxcpp committed:
Merge pull request #567 from the-maldridge/netauth-docs

docs/reference/auth: Add information about netauth support

foxcpp committed:
config: Add EnumMapped, EnumListMapped generic helpers

foxcpp committed:
table: Implement email_with_domains module

foxcpp committed:
table: Implement email_localpart_optional module

foxcpp committed:
Implement auth_map and storage_map at endpoint level

This makes auth_map do what its name implies. Old auth_map in storage
module is deprecated and will be removed in the next release.

foxcpp commented on docs/reference/endpoints/imap.md:
```diff
@@ -36,6 +36,22 @@ tls cert.crt key.key {
+If a list of trusted IP addresses or subnets is provided, only connections
+from those will be trusted.
+
+TLS for the channel between the proxies and maddy can be configured
```
Is `tls` top-level directive unused then?

drdaeman replied:
Oh, good catch! Yes, that's correct, [`inheritGlobal` is `false`](https://github.com/drdaeman/maddy/blob/3910d91be11145312a418560edb72c710301f6e3/internal/endpoint/imap/imap.go#L87) for the `tls` directive in `ProxyProtocol`. In my own setup, I've used different set of certificates and haven't realized global TLS config is a thing. Unless you think it's fine the way it is, I will work on a change that would make global config respected (in a same manner `tls` directive works for IMAP).

foxcpp replied:
In setupListeners in both SMTP and IMAP, created listener is wrapped using tls.NewListener using endpoint-level TLS config before proxy_protocol.NewListener is called. I believe proxy_protocol tls directive can be simply removed.

foxcpp committed:
Merge pull request #566 from arisudesu-forks/loop-extract

Extract mapping lookup from loop

foxcpp committed:
Unbreak tests

foxcpp committed:
auth: Actually check SASL identity argument before using it

Fixes GHSA-4g76-w3xw-2x6w.

foxcpp committed:
Bump go-imap-sql version

Increases open mailbox limit to 100k.

foxcpp committed:
Merge branch '0.6-fixes'

drdaeman commented:
Hey, sorry for a long time to respond - had a few things going on in my life that took all of my time. My bad.

I've rebased my branch and updated the tests, using the recommended 127.0.0.17 address - good consideration, thank you.

I gave it some thought and I'm a bit reluctant to remove the `tls` directive from `proxy_protocol`. Here's my logic:
- Normally, PROXY protocol is used between a load balancer (LB) and individual server (Maddy), and those are typically connected by a trusted network - so, usually one wouldn't use TLS on top of the LB-to-Maddy connection.

    LB ---PROXY[TLS[IMAP]]----> Maddy

  Here, global or IMAP (or SMTP)-level TLS would secure the communications inside the proxied connection, and the connection header would be sent as plaintext. Making `proxy_protocol`'s `tls` inherit the configuration from the top-level `tls` directive would result in odd behavior, where TLS is applied on top of a wrong layer.

- You are absolutely correct that if we'd just remove `tls` from `proxy_protocol` entirely, the issue would be moot.

- However, I believe there is a legit use case where having `tls` would be useful. In case the connection between the LB and Maddy is not trusted (e.g. LB is CloudFlare-like service and the connection is over Internet, not some private LAN/VPC), one may want to put TLS *on top* of PROXY protocol, so no one in position to observe or mess with the traffic between the LB and Maddy would be able to learn or do anything harmful:
  
    LB ---TLS[PROXY[IMAP]]----> Maddy
  
  or possibly even
    
    LB ---TLS_private[PROXY[TLS_public[IMAP]]]----> Maddy
  
  Which is why I'd suggest to keep the directive available, but required to be configured very explicitly, without any inheritance (I've amended docs slightly to be clear about this). It's a low-maintenance thing after all, because all the primitives (`tls2.TLSDirective` and `tls.NewListener`) needed to support it are already there.
  
  I admit, in such scenarios, one would most likely want to also have TLS client certificates for mutual authentication (so Maddy can authenticate the LB as well, possibly relying on PKI trust rather than IP whitelist), but I'd say that's some extra work that is kind of out-of-scope here for the base PROXY protocol support and that it could be done separately at some later point.

I hope this makes sense. If it doesn't, or you feel like that's not a scenario you want to support - let me know and I'll just nuke `tls` directive from there.

Again, sorry for a long silence and thanks!

foxcpp commented:
Thanks for your work!
